File: Algorithm/Java/src/main/java/com/yugutou/charpter1_linklist/level2/topic2_4双指针/GetKthFromEnd.java
package com.yugutou.charpter1_linklist.level2.topic2_4双指针;

/**
 * 寻找链表倒数第K个结点
 */
public class GetKthFromEnd {

    public static void main(String[] args) {
        int[] a = {1, 2, 3, 4, 5};
        ListNode nodeA = initLinkedList(a);
        ListNode node = getKthFromEnd(nodeA,2);
        System.out.println(node.val);
    }

    public static ListNode getKthFromEnd(ListNode head, int k) {
        ListNode fast = head;
        ListNode slow = head;

        while (fast != null && k > 0) {
            fast = fast.next;
            k--;
        }
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }
        return slow;
    }


    private static ListNode initLinkedList(int[] array) {
        ListNode head = null, cur = null;

        for (int i = 0; i < array.length; i++) {
            ListNode newNode = new ListNode(array[i]);
            newNode.next = null;
            if (i == 0) {
                head = newNode;
                cur = head;
            } else {
                cur.next = newNode;
                cur = newNode;
            }
        }
        return head;
    }

    static class ListNode {
        public int val;
        public ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }
}

File: Algorithm/c_or_cpp/charpter1_linklist/level2/topic2_4双指针/get_kth_from_end.cpp
#include <stdlib.h>
#include <stdio.h>

#include <stdio.h>
#include "../../list_node.cpp"

struct ListNode* getKthFromEnd(struct ListNode* head, int k) {
    struct ListNode* fast = head;
    struct ListNode* slow = head;

    while (fast != NULL && k > 0) {
        fast = fast->next;
        k--;
    }

    while (fast != NULL) {
        fast = fast->next;
        slow = slow->next;
    }

    return slow;
}

int main()
{
    struct ListNode *list1 = NULL;

    int a[] = {1, 2, 3, 4, 5};

    list1 = initLink(a, 5);

    struct ListNode *list3 = getKthFromEnd(list1,2);

    // 输出的是4 5,我们主要看第一个是4就可以了，不必纠结后面的结果，想一下为什么
    printList(list3);
    return 0;
}